### PR TITLE
Properly URL decode file:// URIs

### PR DESCRIPTION
**Pull request by sauyon**
I wasn't sure where to put the helper functions so I just dumped them before the function that used them.

### CODE REVIEW

File: icon.c [hunk 1]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
 	return longest > square_size ? square_size/longest : 1.0;
 }
 
+static char hexVal(char digit) {
```
Comment (emersion) [inline icon.c]: Style: use `snake_case`

File: icon.c [hunk 2]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
...
+	assert(isxdigit(digit));
+	if (digit >= 'a') return digit - 'a' + 10;
```
Comment (emersion) [inline icon.c]: Style: always use brackets, even for one-line `if` body

File: icon.c [hunk 3]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
...
+	else if (digit >= 'A') return digit - 'A' + 10;
+	else return digit - '0';
+}
+
+static void urldecode(char *dst, const char *src) {
+	while (*src) {
```
Comment (emersion) [inline icon.c]: Prefer `src[0]` over `*src`

File: icon.c [hunk 4]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
...
+		if ((*src == '%') &&
+				(isxdigit(src[1]) && isxdigit(src[2]))) {
```
Comment (emersion) [inline icon.c]: Style: unnecessary parentheses

File: icon.c [hunk 5]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
...
+			*dst++ = 16*hexVal(src[1]) + hexVal(src[2]);
+			src+=3;
```
Comment (emersion) [inline icon.c]: Style: missing spaces around operator

File: icon.c [hunk 6]
```diff
@@ -49,6 +50,26 @@ static double fit_to_square(int width, int height, int square_size) {
...
+	if (digit >= 'a') return digit - 'a' + 10;
+	else if (digit >= 'A') return digit - 'A' + 10;
+	else return digit - '0';
+}
+
+static void urldecode(char *dst, const char *src) {
+	while (*src) {
+		if ((*src == '%') &&
+				(isxdigit(src[1]) && isxdigit(src[2]))) {
+			*dst++ = 16*hexVal(src[1]) + hexVal(src[2]);
```
Comment (emersion) [inline icon.c]: Please advance `dst` in a separate instruction

File: icon.c [hunk 7]
```diff
@@ -67,8 +88,11 @@ static char *resolve_icon(struct mako_notification *notif) {
 		return strdup(icon_name);
 	}
 	if (strstr(icon_name, "file://") == icon_name) {
-		// Just chop off the scheme.
-		return strdup(icon_name + strlen("file://"));
+		// Chop off the scheme and URL decode
+		char *res = malloc(strlen(icon_name)+1-strlen("file://"));
```
Comment (emersion) [inline icon.c]: Style: missing spaces around `+` and `-`

File: icon.c [hunk 8]
```diff
@@ -67,8 +88,11 @@ static char *resolve_icon(struct mako_notification *notif) {
...
+		if (!res) return res;
```
Comment (emersion) [inline icon.c]: Style: brackets for one-line condition body

**Comment by sauyon** [comment]
Should have looked more carefully at the surrounding code; apologies. Was a quick fix I wrote up at work, and some of the (imo) ugliness from the [StackOverflow solution](https://stackoverflow.com/a/14530993) the code is based on was left over.

Re the fact that it's StackOverflow code I've modified it enough that I'm reasonably sure it doesn't need attribution at this point (and indeed there is no attribution required by default) but if you'd like I can add a comment or something.

**Comment by sauyon** [comment]
Ok, finished my own style audit.

Is there a reason `assert.h` is included twice btw?

**Comment by Emantor** [comment]
FWIW, code contributed before 01.02.2016 is CC-BY-SA which requires attribution. IANAL of course. Also open source licensing is sometimes a mess.

**Comment by sauyon** [comment]
Ah, missed that. I’m still reasonably sure that especially at this point the code is different enough (and/or basic enough) that it should be fine, but obviously IANAL as well.

**Comment by emersion** [comment]
>Is there a reason assert.h is included twice btw?

Ah, this is a mistake

**Pull request merged**